asynciterator.ts

```diff
@@ -1251,6 +1258,119 @@ function destinationFillBuffer<S>(this: InternalSource<S>) {
+}
+
+export class MappingIterator<T, D = T> extends AsyncIterator<D> {
+  private _fn?: Function;
+  private _destroySource: boolean;
+  private onSourceError = (err: Error) => this.emit('error', err);
+  private onSourceReadable = () => this.emit('readable');
+  private onSourceEnd = () => this.close();
+
+  get readable() {
+    return this.source.readable;
+  }
+
+  set readable(readable) {
+    this.source.readable = readable;
+  }
+
+  constructor(
+    protected source: AsyncIterator<T>,
+    private transforms?: ComposedFunction,
+    private upstream: AsyncIterator<any> = source,
+    options: { destroySource?: boolean } = {}
+  ) {
+    // Subscribe the iterator directly upstream rather than the original source to avoid over-subscribing
+    // listeners to the original source
+    super();
+    this._destroySource = options.destroySource !== false;
+    if (upstream.done) {
+      this.close();
+    }
+    else {
+      upstream.on('end', this.onSourceEnd);
+      upstream.on('error', this.onSourceError);
+      upstream.on('readable', this.onSourceReadable);
+    }
+  }
+
+  get fn() {
+    if (!this._fn) {
+      const funcs: Function[] = [];
+      // eslint-disable-next-line prefer-destructuring
+      let transforms: ComposedFunction | undefined = this.transforms!;
+      do
+        funcs.push(transforms.fn);
+      // eslint-disable-next-line no-cond-assign
+      while (transforms = transforms.next);
+
+      const endIndex = funcs.length - 1;
+      this._fn = (item: any) => {
+        // Do not use a for-of loop here, it slows down transformations
+        // by approximately a factor of 2.
+        for (let index = endIndex; index >= 1; index -= 1) {
+          if ((item = funcs[index](item)) === null)
+            return null;
+        }
+        return funcs[0](item);
+      };
+    }
+    return this._fn;
+  }
+
+  read(): D | null {
+    const { source, fn } = this;
+    let item;
+    while ((item = source.read()) !== null) {
+      if ((item = fn(item)) !== null)
+        return item;
+    }
+    return null;
+  }
+
+  map<K>(map: (item: D) => K | null, self?: any): AsyncIterator<K> {
+    return new MappingIterator<T, K>(this.source, { fn: bind(map, self), next: this.transforms }, this);
+  }
+
+  destroy(cause?: Error): void {
+    this.upstream.destroy(cause);
+    super.destroy(cause);
+  }
+
+  public close() {
+    this.upstream.removeListener('end', this.onSourceEnd);
+    this.upstream.removeListener('error', this.onSourceError);
+    this.upstream.removeListener('readable', this.onSourceReadable);
+    if (this._destroySource)
+      this.upstream.destroy();
+    scheduleTask(() => {
+      delete this.source;
+    });
+    super.close();
+  }
+}
+
+export class HeadIterator<T> extends MappingIterator<T> {
+  protected count: number = 0;
+
+  constructor(source: AsyncIterator<T>, protected readonly limit: number) {
+    super(source);
+  }
+
+  read(): T | null {
+    const item = this.source.read();
```

Comment:
Should probably be `super.read()` so all checks (are we not closed?) are still done.